Feature: Add private_key_jwt for OIDC Identity Providers

strehle opened the pull request:
Goal: >>> Allow to omit relyingPartySecret setting <<<

Support private_key_jwt for OIDC servers which provide this, e.g. okta, connect2d, SAP IAM.
[Microsoft Azure AD is not working yet, because of x5t claim. UAA needs to understand and work with X509 for JWT.](https://github.com/cloudfoundry/uaa/issues/2109)

Currently suport OIDC and OAuth2 standards.
Allow to omit relyingPartySecret setting but activate private_key_jwt with jwtClientAuthentication configuration.

If empty, then following OIDC rules

https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication if defined, allow to have own iss and aud claims 

as they are described in RFC 7523.

strehle committed:
review part 1

strehle committed:
Feature: Add private_key_jwt for OIDC Identity Providers

Allow to omit relyingPartySecret setting but activate private_key_jwt
with jwtClientAuthentication configuration.
If empty, then following OIDC rules
https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication
if defined, allow to have own iss and aud claims as they are described in RFC 7523.

strehle committed:
test

strehle committed:
test

strehle committed:
more tests

strehle committed:
more tests

strehle committed:
sonar

strehle commented:
sonar

https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2106

strehle commented:
@torsten-sap tested this with https://help.sap.com/docs/IDENTITY_AUTHENTICATION/6d6d63354d1242d185ab4830fc04feb1/1bdc7291de0e4e529648bfe05856029a.html?locale=en-US

swalchemist commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthentication.java:
```diff
@@ -0,0 +1,62 @@
+    this.keyInfoService = keyInfoService;
+  }
+
+  public String getClientAssetion(OIDCIdentityProviderDefinition config) {
```
Typo - seems that "Assertion" is misspelled in the method name.

swalchemist commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthentication.java:
```diff
@@ -0,0 +1,62 @@
+package org.cloudfoundry.identity.uaa.oauth.jwt;
```
Why no unit test coverage for this class from any of the tests in the server module?

strehle replied:
created test

swalchemist commented on server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/OauthIDPWrapperFactoryBean.java:
```diff
@@ -61,6 +61,16 @@ else if(OIDC10.equalsIgnoreCase(type)) {
                         oidcIdentityProviderDefinition.setPasswordGrantEnabled(idpDefinitionMap.get("passwordGrantEnabled") == null ? false : (boolean) idpDefinitionMap.get("passwordGrantEnabled"));
                         oidcIdentityProviderDefinition.setSetForwardHeader(idpDefinitionMap.get("setForwardHeader") == null ? false : (boolean) idpDefinitionMap.get("passwordGrantEnabled"));
                         oidcIdentityProviderDefinition.setPrompts((List<Prompt>) idpDefinitionMap.get("prompts"));
+                        if (idpDefinitionMap.get("jwtclientAuthentication") != null) {
```
The logic in this method is already too deeply nested. Please extract some methods to simplify the code.

strehle replied:
done

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/authentication/manager/PasswordGrantAuthenticationManagerTest.java:
```diff
@@ -354,7 +360,47 @@ void testOIDCPasswordGrantProviderNoRelyingPartyCredentials() {
+    }
+
+    @Test
+    void testOIDCPasswordGrantProviderJwtClientCredentials() throws MalformedURLException {
```
Any way to simplify the rather large setup for this test?

strehle replied:
refactored this. other tests could also now re-use the mock methods, but in all other tests similar setup is done always from scratch

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feature/issue/1933

strehle committed:
more tests for 100% cov.

strehle committed:
refactored OIDC settings

strehle committed:
sonar cov

strehle committed:
good to have tests

Found an issue.
We want to have jwtClientAuthentication only if true or map is set in configuration, check for correct type during runtime

strehle commented:
@swalchemist Thanks for your review. The label was set in order I wanted to have a review from @torsten-sap  or @tack-sap because this PR solves an issue we have with our setup

During refactoring I fould a bug, but now it should work as expected

Sonar:
https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2106

Expectation is:
We use OIDC provider in our setup of UAA (SAP IAM / Azure ) but we need to get rid of the secrets in the configuration.
OIDC defines private_key_jwt mechanism as replacement for secrets.

strehle commented:
Hi @swalchemist do have any further concerns here ?
